### PR TITLE
fix(richtext): correct paragraph span alignment during text editing

### PR DESCRIPTION
**Pull request by mkeeda**
## Overview
This PR resolves a critical bug where `ParagraphStyle` attributes (such as `BlockquoteKey`) would become misaligned or completely disappear when users performed text editing operations (e.g., character deletion, newline insertion) that overlapped with the styled paragraph boundaries.

## Implementation Details
- **Refactoring & Extraction**: Extracted the paragraph snapping logic from `RichStringBuffer` into a shared `ParagraphSnapping.kt` file.
- **Attribute Separation**: Added `filterParagraphAttributes()` and `filterSpanAttributes()` to `AttributeContainer` to correctly decouple block-level styling from inline styling (e.g., `BoldKey`).
- **Resnap Algorithm**: Implemented `resnapParagraphSpans()`, which separates paragraph and span attributes, calculates the newly snapped paragraph bounds, and safely merges them back together using the `transformSpans` (Sweep-line chunking) algorithm. This prevents overlapping span drops.
- **Integration**: `RichTextState.updateRichString()` now invokes `resnapParagraphSpans` to preserve paragraph boundaries continuously during edits.
- **Testing**: Added rigorous edge-case unit tests in `RichTextStateTest` for deletions, newline insertions, and span attribute preservation under re-snapping conditions.

### CODE REVIEW

File: richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * Spans that contain only [SpanAttributeKey]s are left unchanged.
+ * Paragraph spans whose range becomes invalid (e.g., in empty text) are removed.
+ */
+public fun List<RichSpan>.resnapParagraphSpans(text: String): List<RichSpan> {
+    if (this.isEmpty()) return this
+    if (text.isEmpty()) {
+        return mapNotNull { span ->
+            val sAttrs = span.attributes.filterSpanAttributes()
+            if (sAttrs.isNotEmpty()) {
```
Comment (mkeeda) [inline richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt]: This function is only called from `RichTextState.updateRichString()`, which is `internal` itself. Should the visibility be `internal` rather than `public` to keep it out of the library's public API surface?

File: richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    var resultSpans =
+        mapNotNull { span ->
+            val sAttrs = span.attributes.filterSpanAttributes()
+            if (sAttrs.isNotEmpty()) {
+                span.copy(attributes = sAttrs)
+            } else {
+                null
+            }
+        }
```
Comment (mkeeda) [inline richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt]: The span-attribute filtering logic (lines 29-36 and 40-47) is duplicated between the `text.isEmpty()` early-return branch and the normal path. Consider extracting the common filtering into a local variable before the `if` check to reduce duplication:

```kotlin
val spanOnlySpans = mapNotNull { span ->
    val sAttrs = span.attributes.filterSpanAttributes()
    if (sAttrs.isNotEmpty()) span.copy(attributes = sAttrs) else null
}

if (text.isEmpty()) return spanOnlySpans

var resultSpans = spanOnlySpans
// ... rest of the paragraph re-snapping logic
```

File: richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/AttributeContainer.kt [hunk 3]
```diff
@@ -67,6 +67,26 @@ public class AttributeContainer private constructor(
      */
     public fun containsAny(vararg keys: AttributeKey<*>): Boolean = keys.any { containsKey(it) }
 
+    /**
+     * Returns true if this container contains any [ParagraphAttributeKey].
+     */
+    public fun hasParagraphAttributes(): Boolean =
+        attributes.keys.any { it is ParagraphAttributeKey<*> }
```
Comment (mkeeda) [inline richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/AttributeContainer.kt]: `hasParagraphAttributes()` doesn't appear to be used anywhere in this PR. Is this intended for future use? If so, it might be better to add it in the PR that actually needs it, to avoid unused public API surface.

**Review by mkeeda — COMMENTED** [review]
(AI review)

Overall the approach is solid — extracting `snapToParagraphs` into a shared utility and using `transformSpans` to re-merge paragraph attributes after editing is clean and well-tested. CI is green. ✅

A few suggestions:
1. **Visibility**: `resnapParagraphSpans` could be `internal` since it's only used within the library.
2. **Duplication**: The span-filtering logic in `resnapParagraphSpans` is duplicated between the empty-text branch and the normal path.
3. **Unused API**: `hasParagraphAttributes()` is not referenced anywhere in this PR.

File: richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt (hunk 1 above)
Comment (mkeeda) [reply richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt]: Good catch! I initially tried to change it to `internal`, but since `RichTextState` is located in the `richtext-editor` module and `resnapParagraphSpans` is in the `richtext` module, making it `internal` causes a cross-module compilation error. I have reverted it back to `public`. If we want to hide it from the public API, we might consider introducing an `@InternalRichTextApi` opt-in annotation in a future PR.

File: richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt (hunk 2 above)
Comment (mkeeda) [reply richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt]: Done. Extracted the common filtering logic into `spanOnlySpans` to reduce duplication.

File: richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/AttributeContainer.kt (hunk 3 above)
Comment (mkeeda) [reply richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/AttributeContainer.kt]: Removed `hasParagraphAttributes()` as requested to keep the API surface clean.

File: richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt (hunk 1 above)
Comment (mkeeda) [reply richtext/src/main/kotlin/dev/mkeeda/arranger/richtext/ParagraphSnapping.kt]: Intro `@InternalArrangerApi` and opt-in this annotation in default in entire library modules.
[d26d5fd](https://github.com/mkeeda/arranger/pull/22/commits/d26d5fda750c29c73495270458c618cefe414a63)
